### DISCUSSION TITLE
Is there a way to convert visual mode range into select mode range?

### DISCUSSION THREAD

**Discussion by soomtong** [JetBrains/ideavim#944] [Q&A] [answered]
First of all, thank you for this wonderful plugin provided.

Here is a scenario:

1. Enter visual mode by pressing `v`.
2. Select a range.
3. Change to select mode (yes, it's the idea select mode) or directly input a key, e.g., `"`.

For the case of surrounding a quote:

If it's not supported, `i/I` or `a/A` can trigger this action.

1. Enter visual mode by pressing `v` or `V`.
2. Select a range.
3. Input `A` to change to idea select mode, placing the cursor at the end of the selection range.
4. Alternatively, input `I` to change to idea select mode, placing the cursor at the beginning of the selection range.
5. And `i` or `a` works we expected as like above.

**Comment by citizenmatt** [comment, marked as answer]
It works the same as in Vim - you can use `<C-G>` to switch from Visual to Select mode. See the [help for `v_CTRL-g`](https://vimhelp.org/visual.txt.html#v_CTRL-G).

**Reply by soomtong** [reply]
aha! I didn't know it. that key in my keymap `<c-g>` works idea editor feature.
thanks!

**Reply by huyz** [reply]
Out of curiosity, what's the advantage of switching to Select mode? I can't figure out what it's for (besides letting mouse selection act as in other editors).

**Reply by citizenmatt** [reply]
For normal Vim, I think it's mostly just a preference. Some people prefer selection to work like non-modal editors, where typing will overwrite the selection, rather than invoke a command on it. 

In IdeaVim, it's useful for interacting with IDE features. As mentioned above, in Select mode, you can type something like `"` to have the selected text surrounded by quotes, etc. It's also used as the default mode in IdeaVim when refactoring or expanding Live Templates. For example, when renaming a variable, the IDE will select the current name, ready for editing. Instead of treating this selection as Visual, IdeaVim will recognise that you're refactoring, and switch to Select mode. Since you explicitly invoked a rename refactoring (or similar), it's intuitive to just start typing to change the selected text, rather than have a two step process - invoke the refactoring, then change the mode to edit/overwrite. 

Of course, some people prefer to have more control here. If I want to edit a part of the name while refactoring, then it's more powerful to use Vim commands, and now I've got that two step process again - invoke the refactoring, exit Select mode, then edit. Personally, I overwrite the whole variable more often than just editing it, so prefer to start in Select and just start typing. But you can change this behaviour with the `'idearefactormode'` option, as well as the standard `'keymodel'` and `'selectmode'` options. Check out the [doc on Select mode](https://github.com/JetBrains/ideavim/blob/master/doc/Select-mode.md) for more details and links.

**Reply by huyz** [reply]
Awesome

**Reply by soomtong** [reply]
> I can't figure out what it's for (besides letting mouse selection act as in other editors).

I try do not use mouse as possible :) yes. this is the reason why I'm using vim mode.

you can try it below.
select word and make quote (surround it) → `viw'` this is what I wanted.

Now. I can do it by `viw<c-g>'` my happiness.

**Reply by huyz** [reply]
Nice, although I prefer using vim-surround shortcuts in this case: `ys'` (Or for non-word boundaries: `v<motions>S'`)

I'll be on the lookout for other use cases though

**Reply by soomtong** [reply]
Appreciate! `v<motion>S'` also good for me.
I didn't know capital S.
